// @dynatrace/react-native-plugin/lib/instrumentor/base/Dynatrace.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Dynatrace = void 0;
const Types_1 = require("../model/Types");
const StringUtils_1 = require("./util/StringUtils");
const DynatraceAction_1 = require("./DynatraceAction");
const DynatraceBridge_1 = require("./DynatraceBridge");
const NullAction_1 = require("./NullAction");
const DataCollectionLevel_1 = require("./model/DataCollectionLevel");
const UserPrivacyOptions_1 = require("./UserPrivacyOptions");
const Logger_1 = require("./Logger");
const Configuration_1 = require("./configuration/Configuration");
const ErrorHandler_1 = require("./ErrorHandler");
let counter = 0;
const getCounter = () => counter++;
exports.Dynatrace = {
    start: (configuration) => __awaiter(void 0, void 0, void 0, function* () {
        if (!Configuration_1.Configuration.isConfigurationAvailable()) {
            if (configuration !== undefined) {
                yield DynatraceBridge_1.DynatraceNative.start(configuration);
                if (configuration.reportCrash) {
                    (0, ErrorHandler_1.ErrorHandler)(exports.Dynatrace, Logger_1.Logger).registerErrorHandler();
                }
                Configuration_1.Configuration.setConfiguration(configuration);
                Logger_1.Logger.logDebug('Manual Configuration set: ' + JSON.stringify(configuration));
            }
            else {
                throw new Error('Configuration is empty! Not allowed for Manual Startup!');
            }
            Logger_1.Logger.logInfo('Dynatrace React Native Plugin started!');
        }
    }),
    withMonitoring: (Component, name) => {
        if (Component !== undefined) {
            if (Component.prototype !== undefined && Component.prototype.isReactComponent !== undefined) {
                Component._dtInfo = {
                    type: Types_1.Types.ClassComponent,
                    name,
                };
            }
            else {
                Component._dtInfo = {
                    type: Types_1.Types.FunctionalComponent,
                    name,
                };
            }
        }
        return Component;
    },
    enterAction: (name, platform) => {
        if (Configuration_1.Configuration.isConfigurationAvailable()) {
            const key = getCounter();
            if (!StringUtils_1.StringUtils.isStringNullEmptyOrUndefined(name)) {
                DynatraceBridge_1.DynatraceNative.enterAction(name, key, platform);
                Logger_1.Logger.logDebug(`Enter Action - ${name}`);
                return new DynatraceAction_1.DynatraceAction(key, name);
            }
            else {
                Logger_1.Logger.logDebug('Action Name was empty or null! Action will have no effect.');
                return new NullAction_1.NullAction();
            }
        }
        else {
            Logger_1.Logger.logInfo("React Native plugin has not been started yet! Action can't be created!");
            return new NullAction_1.NullAction();
        }
    },
    reportError: (errorName, errorCode, platform) => {
        if (Configuration_1.Configuration.isConfigurationAvailable()) {
            DynatraceBridge_1.DynatraceNative.reportErrorWithoutStacktrace(errorName, errorCode, platform);
        }
        else {
            Logger_1.Logger.logInfo('React Native plugin has not been started yet! Error will not be reported!');
        }
    },
    reportErrorFromHandler: (isFatal, errorName, reason, stacktrace, platform) => {
        if (!StringUtils_1.StringUtils.isStringNullEmptyOrUndefined(reason)) {
            const reasonNewLineIndex = reason.indexOf('\n');
            if (reasonNewLineIndex !== -1) {
                reason = reason.substring(0, reasonNewLineIndex);
            }
        }
        if (isFatal) {
            DynatraceBridge_1.DynatraceNative.reportCrash(errorName, reason, stacktrace, true, platform);
        }
        else {
            DynatraceBridge_1.DynatraceNative.reportError(errorName, '-', reason, stacktrace, platform);
        }
    },
    reportErrorWithStacktrace: (errorName, reason, stacktrace, platform) => {
        if (Configuration_1.Configuration.isConfigurationAvailable()) {
            if (!StringUtils_1.StringUtils.isStringNullEmptyOrUndefined(reason)) {
                const reasonNewLineIndex = reason.indexOf('\n');
                if (reasonNewLineIndex !== -1) {
                    reason = reason.substring(0, reasonNewLineIndex);
                }
            }
            DynatraceBridge_1.DynatraceNative.reportError(errorName, '-', reason, stacktrace, platform);
        }
        else {
            Logger_1.Logger.logInfo('React Native plugin has not been started yet! Error will not be reported!');
        }
    },
    reportErrorStacktrace: (errorName, errorValue, reason, stacktrace, platform) => {
        if (Configuration_1.Configuration.isConfigurationAvailable()) {
            if (!StringUtils_1.StringUtils.isStringNullEmptyOrUndefined(reason)) {
                const reasonNewLineIndex = reason.indexOf('\n');
                if (reasonNewLineIndex !== -1) {
                    reason = reason.substring(0, reasonNewLineIndex);
                }
            }
            DynatraceBridge_1.DynatraceNative.reportError(errorName, errorValue, reason, stacktrace, platform);
        }
        else {
            Logger_1.Logger.logInfo('React Native plugin has not been started yet! Error will not be reported!');
        }
    },
    reportCrash: (crashName, reason, stacktrace, platform) => {
        if (Configuration_1.Configuration.isConfigurationAvailable()) {
            if (!StringUtils_1.StringUtils.isStringNullEmptyOrUndefined(crashName)) {
                if (!StringUtils_1.StringUtils.isStringNullEmptyOrUndefined(reason)) {
                    const reasonNewLineIndex = reason.indexOf('\n');
                    if (reasonNewLineIndex !== -1) {
                        reason = reason.substring(0, reasonNewLineIndex);
                    }
                }
                DynatraceBridge_1.DynatraceNative.reportCrash(crashName, reason, stacktrace, false, platform);
            }
        }
        else {
            Logger_1.Logger.logInfo('React Native plugin has not been started yet! Crash will not be reported!');
        }
    },
    reportCrashWithException: (crashName, crash, platform) => {
        if (Configuration_1.Configuration.isConfigurationAvailable()) {
            if (!StringUtils_1.StringUtils.isStringNullEmptyOrUndefined(crashName) &&
                !StringUtils_1.StringUtils.isStringNullEmptyOrUndefined(crash.stack)) {
                let crashReason = crash.message;
                if (!StringUtils_1.StringUtils.isStringNullEmptyOrUndefined(crashReason)) {
                    const reasonNewLineIndex = crashReason.indexOf('\n');
                    if (reasonNewLineIndex !== -1) {
                        crashReason = crashReason.substring(0, reasonNewLineIndex);
                    }
                }
                DynatraceBridge_1.DynatraceNative.reportCrash(crashName, crashReason, crash.stack, true, platform);
            }
        }
        else {
            Logger_1.Logger.logInfo('React Native plugin has not been started yet! Crash will not be reported!');
        }
    },
    sendEvent: (name, attributes = {}, platform) => {
        if (Configuration_1.Configuration.isConfigurationAvailable()) {
            if (StringUtils_1.StringUtils.isStringNullEmptyOrUndefined(name)) {
                Logger_1.Logger.logDebug('sendEvent: Name must be a non empty string');
                return;
            }
            if (typeof attributes !== 'object' ||
                attributes === null ||
                Array.isArray(attributes)) {
                Logger_1.Logger.logDebug('sendEvent: Payload toplevel must be an object');
                return;
            }
            const internalAttributes = Object.assign({ 'dt.agent.flavor': 'react_native' }, attributes);
            Logger_1.Logger.logDebug(`sendEvent(${name}, ${internalAttributes})`);
            DynatraceBridge_1.DynatraceNative.sendEvent(name, internalAttributes, platform);
        }
        else {
            Logger_1.Logger.logInfo('React Native plugin has not been started yet! Event will not be reported!');
        }
    },
    sendBizEvent: (type, attributes = {}, platform) => {
        if (Configuration_1.Configuration.isConfigurationAvailable()) {
            if (StringUtils_1.StringUtils.isStringNullEmptyOrUndefined(type)) {
                Logger_1.Logger.logDebug('sendBizEvent: Type must be a non empty string');
                return;
            }
            if (typeof attributes !== 'object' ||
                attributes === null ||
                Array.isArray(attributes)) {
                Logger_1.Logger.logDebug('sendBizEvent: Payload toplevel must be an object');
                return;
            }
            const internalAttributes = Object.assign({ 'dt.agent.flavor': 'react_native' }, attributes);
            Logger_1.Logger.logDebug(`sendBizEvent(${type}, ${internalAttributes})`);
            DynatraceBridge_1.DynatraceNative.sendBizEvent(type, internalAttributes, platform);
        }
        else {
            Logger_1.Logger.logInfo('React Native plugin has not been started yet! Biz Event will not be reported!');
        }
    },
    endSession: (platform) => {
        if (Configuration_1.Configuration.isConfigurationAvailable()) {
            DynatraceBridge_1.DynatraceNative.endVisit(platform);
        }
        else {
            Logger_1.Logger.logInfo('React Native plugin has not been started yet! Session will not be ended!');
        }
    },
    identifyUser: (user, platform) => {
        if (Configuration_1.Configuration.isConfigurationAvailable()) {
            DynatraceBridge_1.DynatraceNative.identifyUser(user, platform);
        }
        else {
            Logger_1.Logger.logInfo('React Native plugin has not been started yet! Session will not be tagged!');
        }
    },
    setGPSLocation: (latitude, longitude, platform) => {
        if (Configuration_1.Configuration.isConfigurationAvailable()) {
            DynatraceBridge_1.DynatraceNative.setGPSLocation(latitude, longitude, platform);
        }
        else {
            Logger_1.Logger.logInfo('React Native plugin has not been started yet! GPS Position will not be set!');
        }
    },
    isCrashReportingOptedIn: (platform) => __awaiter(void 0, void 0, void 0, function* () {
        if (Configuration_1.Configuration.isConfigurationAvailable()) {
            return DynatraceBridge_1.DynatraceNative.isCrashReportingOptedIn(platform);
        }
        else {
            Logger_1.Logger.logInfo('React Native plugin has not been started yet! Crash reporting value not available!');
            return false;
        }
    }),
    setCrashReportingOptedIn: (crashReporting, platform) => {
        if (Configuration_1.Configuration.isConfigurationAvailable()) {
            DynatraceBridge_1.DynatraceNative.setCrashReportingOptedIn(crashReporting, platform);
        }
        else {
            Logger_1.Logger.logInfo('React Native plugin has not been started yet! Crash reporting value will not be set!');
        }
    },
    getDataCollectionLevel: (platform) => __awaiter(void 0, void 0, void 0, function* () {
        if (Configuration_1.Configuration.isConfigurationAvailable()) {
            return DynatraceBridge_1.DynatraceNative.getDataCollectionLevel(platform);
        }
        else {
            Logger_1.Logger.logInfo('React Native plugin has not been started yet! Data Collection level not available!');
            return DataCollectionLevel_1.DataCollectionLevel.Off;
        }
    }),
    setDataCollectionLevel: (dataCollectionLevel, platform) => {
        if (Configuration_1.Configuration.isConfigurationAvailable()) {
            DynatraceBridge_1.DynatraceNative.setDataCollectionLevel(dataCollectionLevel, platform);
        }
        else {
            Logger_1.Logger.logInfo('React Native plugin has not been started yet! Data Collection level will not be set!');
        }
    },
    getUserPrivacyOptions: (platform) => __awaiter(void 0, void 0, void 0, function* () {
        if (Configuration_1.Configuration.isConfigurationAvailable()) {
            const options = yield DynatraceBridge_1.DynatraceNative.getUserPrivacyOptions(platform);
            const currentOptions = new UserPrivacyOptions_1.UserPrivacyOptions(options.dataCollectionLevel, options.crashReportingOptedIn);
            return currentOptions;
        }
        else {
            Logger_1.Logger.logInfo('React Native plugin has not been started yet! User Privacy Options are not available!');
            return new UserPrivacyOptions_1.UserPrivacyOptions(DataCollectionLevel_1.DataCollectionLevel.Off, false);
        }
    }),
    applyUserPrivacyOptions: (userPrivacyOptions, platform) => {
        if (Configuration_1.Configuration.isConfigurationAvailable()) {
            if ((userPrivacyOptions.crashReportingOptedIn === true ||
                userPrivacyOptions.crashReportingOptedIn === false) &&
                userPrivacyOptions.dataCollectionLevel in DataCollectionLevel_1.DataCollectionLevel) {
                DynatraceBridge_1.DynatraceNative.applyUserPrivacyOptions(userPrivacyOptions, platform);
            }
        }
        else {
            Logger_1.Logger.logInfo("React Native plugin has not been started yet! User Privacy Options can't be applied!");
        }
    },
    flushEvents: (platform) => {
        if (Configuration_1.Configuration.isConfigurationAvailable()) {
            DynatraceBridge_1.DynatraceNative.flushEvents(platform);
        }
        else {
            Logger_1.Logger.logInfo('React Native plugin has not been started yet! Flushing Event not possible!');
        }
    },
    setBeaconHeaders: (headers, platform) => {
        if (Configuration_1.Configuration.isConfigurationAvailable()) {
            if (headers && headers.size > 0) {
                const obj = {};
                headers.forEach((value, key) => (obj[key] = value));
                DynatraceBridge_1.DynatraceNative.setBeaconHeaders(obj, platform);
            }
            else if (!headers || typeof headers !== 'undefined') {
                DynatraceBridge_1.DynatraceNative.setBeaconHeaders(null, platform);
            }
        }
        else {
            Logger_1.Logger.logInfo('React Native plugin has not been started yet! Setting Beacon Headers is not possible!');
        }
    },
};
