// @dynatrace/react-native-plugin/lib/instrumentor/base/configuration/ManualStartupConfiguration.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ManualStartupConfiguration = void 0;
const ConfigurationDefaults_1 = require("./ConfigurationDefaults");
class ManualStartupConfiguration {
    constructor(beaconUrl, applicationId, reportCrash, logLevel, lifecycleUpdate, userOptIn, actionNamePrivacy, bundleName) {
        this.reportCrash = ConfigurationDefaults_1.DEFAULT_REPORT_CRASH;
        this.userOptIn = ConfigurationDefaults_1.DEFAULT_USER_OPT_IN;
        this.logLevel = ConfigurationDefaults_1.DEFAULT_LOGLEVEL;
        this.lifecycleUpdate = ConfigurationDefaults_1.DEFAULT_LIFECYCLE_UPDATE;
        this.actionNamePrivacy = ConfigurationDefaults_1.DEFAULT_ACTION_NAME_PRIVACY;
        this.bundleName = ConfigurationDefaults_1.DEFAULT_BUNDLE_NAME;
        if (!applicationId || !beaconUrl) {
            throw new Error('Dynatrace Startup Configuration is not correct! Missing applicationId and beaconUrl!');
        }
        this.applicationId = applicationId;
        this.beaconUrl = beaconUrl;
        if (reportCrash != null) {
            this.reportCrash = reportCrash;
        }
        if (userOptIn != null) {
            this.userOptIn = userOptIn;
        }
        if (logLevel != null) {
            this.logLevel = logLevel;
        }
        if (lifecycleUpdate != null) {
            this.lifecycleUpdate = lifecycleUpdate;
        }
        if (actionNamePrivacy != null) {
            this.actionNamePrivacy = actionNamePrivacy;
        }
        if (bundleName != null) {
            this.bundleName = bundleName;
        }
    }
}
exports.ManualStartupConfiguration = ManualStartupConfiguration;
